File: android/src/main/kotlin/com/anhlp/barcode_scanner/mappers/BarcodeMapper.kt
package com.anhlp.barcode_scanner.mappers

import com.anhlp.barcode_scanner.Protos

object BarcodeMapper {
    fun toBarcodes(barcodes: List<com.google.mlkit.vision.barcode.common.Barcode>): List<Protos.BarcodeResult> {
        return barcodes.map { toBarcode(it) }
    }

    private fun toBarcode(barcode: com.google.mlkit.vision.barcode.common.Barcode): Protos.BarcodeResult {
        val format = BarcodeFormatMapper.toBarcodeFormat(barcode.format)
        val boundingBox = Protos.Rect.newBuilder()
            .setLeft(barcode.boundingBox?.left?.toFloat() ?: 0f)
            .setTop(barcode.boundingBox?.top?.toFloat() ?: 0f)
            .setRight(barcode.boundingBox?.right?.toFloat() ?: 0f)
            .setBottom(barcode.boundingBox?.bottom?.toFloat() ?: 0f)
            .build()

        val builder = Protos.BarcodeResult.newBuilder()
        builder.setFormat(format)
        builder.setRawValue(barcode.rawValue)
        builder.setBoundingBox(boundingBox)
        for (e in barcode.cornerPoints!!) {
            val cornerPointsBuilder = Protos.Point.newBuilder()
            cornerPointsBuilder.setX(e.x.toFloat())
            cornerPointsBuilder.setY(e.y.toFloat())
            builder.addCornerPoints(cornerPointsBuilder.build())
        }
        builder.setTimestamp(System.currentTimeMillis() / 1000)
        return builder.build()
    }

    fun toBarcode(barcode: com.google.zxing.Result): Protos.BarcodeResult {
        val format = BarcodeFormatMapper.toBarcodeFormat(barcode.barcodeFormat)
        val boundingBox =
            barcode.resultPoints?.filterNotNull()?.takeIf { it.size >= 4 }?.let { points ->
                val xs = points.map { it.x }
                val ys = points.map { it.y }
                Protos.Rect.newBuilder()
                    .setLeft(xs.minOrNull() ?: 0f)
                    .setTop(ys.minOrNull() ?: 0f)
                    .setRight(xs.maxOrNull() ?: 0f)
                    .setBottom(ys.maxOrNull() ?: 0f)
                    .build()
            }

        val builder = Protos.BarcodeResult.newBuilder()
        builder.setFormat(format)
        builder.setRawValue(barcode.text)
        builder.setBoundingBox(boundingBox)
        for (e in barcode.resultPoints) {
            val cornerPointsBuilder = Protos.Point.newBuilder()
            cornerPointsBuilder.setX(e.x)
            cornerPointsBuilder.setY(e.y)
            builder.addCornerPoints(cornerPointsBuilder.build())
        }
        builder.setTimestamp(System.currentTimeMillis() / 1000)
        return builder.build()
    }
}

File: ios/Classes/Mappers/BarcodeMapper.swift
import Vision

struct BarcodeMapper {
    static func toBarcodes(from observations: [VNBarcodeObservation]) -> [BarcodeResult] {
        return observations.map(toBarcode(from:))
    }
    
    static func toBarcode(from observation: VNBarcodeObservation) -> BarcodeResult {
        let format = BarcodeFormatMapper.toBarcodeFormat(from: observation.symbology)
        let rect = CGRect(
            x: observation.boundingBox.origin.y,
            y: observation.boundingBox.origin.x,
            width: observation.boundingBox.height,
            height: observation.boundingBox.width
        )
        let boundingBox = Rect.with {
            $0.left = Float(rect.origin.x)
            $0.top = Float(rect.origin.y)
            $0.right = Float(rect.origin.x + rect.width)
            $0.bottom = Float(rect.origin.y + rect.height)
        }
        
        let barcode = BarcodeResult.with {
            $0.format = format
            $0.rawValue = observation.payloadStringValue ?? ""
            if #available(iOS 17.0, *) {
                $0.rawBytes = observation.payloadData ?? Data()
            }
            $0.boundingBox = boundingBox
            $0.cornerPoints = [
                Point.with {
                    $0.x = Float(observation.topLeft.x)
                    $0.y = Float(observation.topLeft.y)
                },
                Point.with {
                    $0.x = Float(observation.topRight.x)
                    $0.y = Float(observation.topRight.y)
                },
                Point.with {
                    $0.x = Float(observation.bottomRight.x)
                    $0.y = Float(observation.bottomRight.y)
                },
                Point.with {
                    $0.x = Float(observation.bottomLeft.x)
                    $0.y = Float(observation.bottomLeft.y)
                }
            ]
            $0.timestamp = Int64(Date().timeIntervalSince1970)
        }
        
        return barcode
    }
}
